src/scala/org/pantsbuild/zinc/compiler/Main.scala

```diff
@@ -187,18 +197,87 @@ object Main {
     Settings.SettingsParser
       .parse(preprocessArgs(context.getArgs), Settings()) match {
       case Some(settings) =>
+        val workingDirectory = new File(context.getWorkingDirectory)
         mainImpl(
-          settings.withAbsolutePaths(new File(context.getWorkingDirectory)),
+          settings.withAbsolutePaths(workingDirectory),
           startTime,
-          n => context.exit(n))
+          n => context.exit(n),
+          workingDirectory
+        )
       case None => {
         println("See zinc-compiler --help for information about options")
         context.exit(1)
       }
     }
   }
 
-  def mainImpl(settings: Settings, startTime: Long, exit: Int => Unit): Unit = {
+  def toLsp(severity: Severity): DiagnosticSeverity = {
+    severity match {
+      case Severity.Error => DiagnosticSeverity.Error
+      case Severity.Warn  => DiagnosticSeverity.Warning
+      case Severity.Info  => DiagnosticSeverity.Information
+      // Note: DiagnosticSeverity also has a Hint level, but we won't use it here as xsbti doesn't go that far
+    }
+  }
+  def toZeroBased(x: Int): Int = {
+    x - 1
+  }
+  def toLsp(problems: List[Problem],
+            workingDirectory: File): List[PublishDiagnosticsParams] = {
+    problems
+      .groupBy(problem => problem.position.sourcePath)
+      .map {
+        case (file, problems) => {
+          // By using a relative path here as opposed to an actual URI, we diverge slightly from
+          // the precise definition of the Language Server Protocol; but this is necessary if we
+          // want to be able to cache this data in heterogeneous environments.
+          val uri: String = "buildroot://" + file.asScala
```

Comment:
nit: If we `getOrElse("")`, this signals to me that the field itself is optional in whatever data we output, and would ask why `uri` is not modeled as an `Option`. If this is because the LSP protocol doesn't allow an option, I'd add a comment saying why we don't care about an empty URI.